File: backend/services/evals/narrativeChecks.js
const OPENING_WINDOW = 200;
const MOVEMENT_PREFIXES = ['camin', 'avanz', 'muev', 'despert', 'levant', 'sigu', 'continu', 'march', 'and', 'part', 'recorr', 'cruzu', 'desplaz'];

function normalize(text = '') {
  return text
    .toLowerCase()
    .normalize('NFD')
    .replace(/[\u0300-\u036f]/g, '');
}

function startsWithAranathMovement(narrative, characterName = 'Aranath') {
  const norm = normalize(narrative);
  const firstSentence = norm.split('.', 1)[0] || '';
  const name = normalize(characterName);
  const words = firstSentence.trim().split(' ').filter(Boolean);
  const startsWithName = words[0] === name;
  const hasWalkingVerb = words.some(w => MOVEMENT_PREFIXES.some(p => w.startsWith(p)));
  return { startsWithName, hasWalkingVerb, firstSentence: firstSentence.trim() };
}

export function checkOpening(narrative, characterName = 'Aranath') {
  console.log('[check:opening] checking opening against', characterName);
  const { startsWithName, hasWalkingVerb, firstSentence } = startsWithAranathMovement(narrative, characterName);
  const ok = !startsWithName && !hasWalkingVerb;
  const result = { name: 'opening_no_aranath_walking', ok, details: { firstSentence } };
  if (!ok) {
    result.reason = startsWithName
      ? 'the first sentence starts with the character name'
      : 'the first words contain a movement/waking verb';
  }
  console.log('[check:opening]', ok ? 'PASS' : 'FAIL');
  return result;
}

export function checkBannedPhrases(narrative, bannedPhrases = []) {
  console.log('[check:banned] checking', bannedPhrases.length, 'banned phrases');
  const text = normalize(narrative);
  const found = [];
  for (const phrase of bannedPhrases) {
    const p = normalize(phrase);
    if (p && text.includes(p)) {
      found.push(phrase);
    }
  }
  const ok = found.length === 0;
  const result = { name: 'banned_phrases', ok, details: { found } };
  if (!ok) result.reason = 'found banned phrase(s): ' + found.join('; ');
  console.log('[check:banned]', ok ? 'PASS' : 'FAIL', found);
  return result;
}

export function checkPromptQuotes(narrative, promptText = '') {
  console.log('[check:prompt_quote] scanning for copied prompt phrases');
  const text = normalize(narrative);
  const normPrompt = normalize(promptText);
  const candidates = normPrompt
    .split('.')
    .map(s => s.trim())
    .filter(s => s.length >= 40);
  const found = [];
  for (const phrase of candidates) {
    if (text.includes(phrase)) {
      found.push(phrase.slice(0, 80));
    }
  }
  const ok = found.length === 0;
  const result = { name: 'prompt_quote', ok, details: { found } };
  if (!ok) result.reason = 'narrative copies ' + found.length + ' prompt phrase(s)';
  console.log('[check:prompt_quote]', ok ? 'PASS' : 'FAIL');
  return result;
}

export function checkEncountersPresented(narrative, encounters = []) {
  console.log('[check:encounters] checking', encounters.length, 'encounters');
  const text = normalize(narrative);
  const missing = [];
  for (const encounter of encounters) {
    const name = encounter?.entity?.name;
    if (!name) continue;
    const tokens = name
      .split('(')[0]
      .trim()
      .toLowerCase()
      .split(' ')
      .flatMap(t => t.split('-'))
      .filter(t => t.length > 2);
    const present = tokens.some(t => text.includes(t));
    if (!present) missing.push(name);
  }
  const ok = missing.length === 0;
  const result = { name: 'encounters_presented', ok, details: { missing } };
  if (!ok) result.reason = 'missing encounter(s): ' + missing.join(', ');
  console.log('[check:encounters]', ok ? 'PASS' : 'FAIL', missing);
  return result;
}

export function checkSceneryInventory(narrative) {
  console.log('[check:scenery_inventory] checking first 500 chars for inventory lists');
  const firstChunk = narrative.slice(0, 500).toLowerCase();
  let count = 0;
  let idx = firstChunk.indexOf('había');
  while (idx !== -1) { count++; idx = firstChunk.indexOf('había', idx + 1); }
  let noCount = 0;
  idx = firstChunk.indexOf('no había');
  while (idx !== -1) { noCount++; idx = firstChunk.indexOf('no había', idx + 1); }
  const total = count + noCount;
  const ok = total <= 2;
  const result = { name: 'scenery_inventory', ok, details: { habia_count: total } };
  if (!ok) result.reason = 'first paragraph uses ' + total + ' había/no había (inventory-style)';
  console.log('[check:scenery_inventory]', ok ? 'PASS' : 'FAIL', total);
  return result;
}

File: story-engine/app/evals/narrative_checks.py
# ============================================================================
# Narrative quality checks
# ----------------------------------------------------------------------------
# Port of backend/domains/story/services/evals/narrativeChecks.js.
# ============================================================================
import unicodedata

OPENING_WINDOW = 200
MOVEMENT_PREFIXES = ['camin', 'avanz', 'muev', 'despert', 'levant', 'sigu', 'continu', 'march', 'and', 'part', 'recorr', 'cruzu', 'desplaz']


def _normalize(text=''):
    text = text or ''
    normalized = unicodedata.normalize('NFD', text.lower())
    return ''.join(c for c in normalized if unicodedata.category(c) != 'Mn')


def _starts_with_aranath_movement(narrative, character_name='Aranath'):
    norm = _normalize(narrative)
    first_sentence = norm.split('.', 1)[0] if norm else ''
    name = _normalize(character_name)
    words = [w for w in first_sentence.strip().split(' ') if w]
    starts_with_name = bool(words) and words[0] == name
    has_walking_verb = any(any(w.startswith(p) for p in MOVEMENT_PREFIXES) for w in words)
    return starts_with_name, has_walking_verb, first_sentence.strip()


def check_opening(narrative, character_name='Aranath'):
    print('[check:opening] checking opening against', character_name)
    starts_with_name, has_walking_verb, first_sentence = _starts_with_aranath_movement(narrative, character_name)
    ok = not starts_with_name and not has_walking_verb
    result = {'name': 'opening_no_aranath_walking', 'ok': ok, 'details': {'firstSentence': first_sentence}}
    if not ok:
        result['reason'] = (
            'the first sentence starts with the character name' if starts_with_name
            else 'the first words contain a movement/waking verb'
        )
    print('[check:opening]', 'PASS' if ok else 'FAIL')
    return result


def check_banned_phrases(narrative, banned_phrases=None):
    banned_phrases = banned_phrases or []
    print('[check:banned] checking', len(banned_phrases), 'banned phrases')
    text = _normalize(narrative)
    found = []
    for phrase in banned_phrases:
        p = _normalize(phrase)
        if p and p in text:
            found.append(phrase)
    ok = len(found) == 0
    result = {'name': 'banned_phrases', 'ok': ok, 'details': {'found': found}}
    if not ok:
        result['reason'] = 'found banned phrase(s): ' + '; '.join(found)
    print('[check:banned]', 'PASS' if ok else 'FAIL', found)
    return result


def check_prompt_quotes(narrative, prompt_text=''):
    print('[check:prompt_quote] scanning for copied prompt phrases')
    text = _normalize(narrative)
    norm_prompt = _normalize(prompt_text)
    candidates = [s.strip() for s in norm_prompt.split('.') if len(s.strip()) >= 40]
    found = []
    for phrase in candidates:
        if phrase in text:
            found.append(phrase[:80])
    ok = len(found) == 0
    result = {'name': 'prompt_quote', 'ok': ok, 'details': {'found': found}}
    if not ok:
        result['reason'] = f'narrative copies {len(found)} prompt phrase(s)'
    print('[check:prompt_quote]', 'PASS' if ok else 'FAIL')
    return result


def check_encounters_presented(narrative, encounters=None):
    encounters = encounters or []
    print('[check:encounters] checking', len(encounters), 'encounters')
    text = _normalize(narrative)
    missing = []
    for encounter in encounters:
        name = (encounter.get('entity') or {}).get('name')
        if not name:
            continue
        tokens = []
        for word in name.split('(')[0].strip().lower().split(' '):
            tokens.extend(word.split('-'))
        tokens = [t for t in tokens if len(t) > 2]
        present = any(t in text for t in tokens)
        if not present:
            missing.append(name)
    ok = len(missing) == 0
    result = {'name': 'encounters_presented', 'ok': ok, 'details': {'missing': missing}}
    if not ok:
        result['reason'] = 'missing encounter(s): ' + ', '.join(missing)
    print('[check:encounters]', 'PASS' if ok else 'FAIL', missing)
    return result


def check_scenery_inventory(narrative):
    print('[check:scenery_inventory] checking first 500 chars for inventory lists')
    first_chunk = (narrative or '')[:500].lower()

    def count_occurrences(needle):
        count = 0
        idx = first_chunk.find(needle)
        while idx != -1:
            count += 1
            idx = first_chunk.find(needle, idx + 1)
        return count

    count = count_occurrences('había')
    no_count = count_occurrences('no había')
    total = count + no_count
    ok = total <= 2
    result = {'name': 'scenery_inventory', 'ok': ok, 'details': {'habia_count': total}}
    if not ok:
        result['reason'] = f'first paragraph uses {total} había/no había (inventory-style)'
    print('[check:scenery_inventory]', 'PASS' if ok else 'FAIL', total)
    return result
